xarray/coding/variables.py

```diff
@@ -219,7 +247,9 @@ def encode(self, variable, name=None):
         dims, data, attrs, encoding = unpack_for_encoding(variable)
 
         if 'scale_factor' in encoding or 'add_offset' in encoding:
-            dtype = _choose_float_dtype(data.dtype, 'add_offset' in encoding)
+            scale_factor = pop_to(attrs, encoding, 'scale_factor', name=name)
+            add_offset = pop_to(attrs, encoding, 'add_offset', name=name)
+            dtype = _choose_float_dtype(data.dtype, scale_factor, add_offset)
             data = data.astype(dtype=dtype, copy=True)
             if 'add_offset' in encoding:
```

Comment:
This line (and the line below for `scale_factor`) needs to be updated to use the `add_offset` variable if it isn't `None` rather than checking for it in `encoding` again. (`pop_to` removes an item from the dict in which it's found.)